ejector/ejector.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/Layr-Labs/eigensdk-go/logging"
+)
+
+// Ejector is responsible for periodically evaluating validators and deciding which ones to eject.
+type Ejector struct {
+	ctx    context.Context
+	logger logging.Logger
+
+	// Responsible for executing ejections and managing the ejection lifecycle.
+	ejectionManager *ThreadedEjectionManager
+
+	// Used for looking up signing rates for V1.
+	// TODO(cody.littley): remove this after V1 sunset
+	signingRateLookupV1 SigningRateLookup
+
+	// Used for looking up signing rates for V2.
+	signingRateLookupV2 SigningRateLookup
+
+	// The freqeuency with which to evaluate validators for ejection.
+	period time.Duration
+
+	// Defines the time window over which to evaluate signing metrics when deciding whether to eject a validator.
+	ejectionCriteriaTimeWindow time.Duration
+
+	// Used to convert validator IDs to validator addresses.
+	validatorIDToAddressCache eth.ValidatorIDToAddressConverter
+
+	// Used to look up the latest reference number.
+	referenceBlockProvider eth.ReferenceBlockProvider
+
+	// Used to fetch operator state.
+	operatorStateCache operatorstate.OperatorStateCache
+}
+
+// NewEjector creates a new Ejector.
+func NewEjector(
+	ctx context.Context,
+	logger logging.Logger,
+	config *EjectorConfig,
+	ejectionManager *ThreadedEjectionManager,
+	signingRateLookupV1 SigningRateLookup,
+	signingRateLookupV2 SigningRateLookup,
+	validatorIDToAddressCache eth.ValidatorIDToAddressConverter,
+) *Ejector {
+	e := &Ejector{
+		ctx:                        ctx,
+		logger:                     logger,
+		ejectionManager:            ejectionManager,
+		signingRateLookupV1:        signingRateLookupV1,
+		signingRateLookupV2:        signingRateLookupV2,
+		period:                     config.EjectionPeriod,
+		ejectionCriteriaTimeWindow: config.EjectionCriteriaTimeWindow,
+		validatorIDToAddressCache:  validatorIDToAddressCache,
+	}
+
+	go e.mainLoop()
+
+	return e
+}
+
+// The main loop periodically evaluates validators for ejection.
+func (e *Ejector) mainLoop() {
+	ticker := time.NewTicker(e.period)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-e.ctx.Done():
+			e.logger.Info("ejector shutting down")
+			return
+		case <-ticker.C:
+			e.evaluateValidators()
+		}
+	}
+}
+
+// evaluateValidators looks up signing rates and evaluates which validators should be ejected.
+func (e *Ejector) evaluateValidators() {
+
+	v1SigningRates, err := e.signingRateLookupV1.GetSigningRates(
+		e.ejectionCriteriaTimeWindow,
+		nil, // all quorums
+		ProtocolVersionV1,
+		true, // omit perfect signers if possible
+	)
+	if err != nil {
+		e.logger.Error("error looking up v1 signing rates", "error", err)
+		return
+	}
+
+	v2SigningRates, err := e.signingRateLookupV2.GetSigningRates(
+		e.ejectionCriteriaTimeWindow,
+		nil, // all quorums
+		ProtocolVersionV2,
+		true, // omit perfect signers if possible
+	)
+	if err != nil {
+		e.logger.Error("error looking up v2 signing rates", "error", err)
```

Comment:
nit: it's a bit strange to have no return value for `evaluateValidators`, and log all errors internally. Why not just return an error, and log it in the main loop if non-nil?

Same comment applies to `evaluateValidator`

Reply:
fixed both